op-node/rollup/status/l1_head_buffer.go

```diff
@@ -0,0 +1,59 @@
+package status
+
+import (
+	"sync"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+// l1HeadBuffer is a thread-safe cache for L1 block references, which contains a series blocks with a valid chain of parent hashes.
+type l1HeadBuffer struct {
+	rb             *ringbuffer[eth.L1BlockRef]
+	minBlockNumber uint64
+	mu             sync.RWMutex
+}
+
+func newL1HeadBuffer(size int) *l1HeadBuffer {
+	return &l1HeadBuffer{rb: newRingBuffer[eth.L1BlockRef](size)}
+}
+
+// Get returns the L1 block reference for the given block number, if it exists in the cache.
+func (lhb *l1HeadBuffer) Get(num uint64) (eth.L1BlockRef, bool) {
+	lhb.mu.RLock()
+	defer lhb.mu.RUnlock()
+
+	return lhb.rb.Get(int(num - lhb.minBlockNumber))
+}
+
+// Insert inserts a new L1 block reference into the cache, and removes any entries that are invalidated by a reorg.
+// If the parent hash of the new head doesn't match the hash of the previous head, all entries after the new head are removed
+// as the chain cannot be validated.
+func (lhb *l1HeadBuffer) Insert(l1Head eth.L1BlockRef) {
+	lhb.mu.Lock()
+	defer lhb.mu.Unlock()
+
+	// First, check if the L1 head is in the cache.
+	// If the hash doesn't match the one in the cache, we have a reorg and need to remove all entries after the new head.
+	if ref, ok := lhb.Get(l1Head.Number); ok {
+		if ref.Hash != l1Head.Hash {
+			// Reorg detected, invalidate all entries after the new head.
+			for {
+				ref, ok := lhb.rb.Pop()
+				if !ok || ref.Number == l1Head.Number {
+					break
+				}
+			}
+			lhb.rb.Push(l1Head)
```

Comment:
I don't think this is sufficient because it isn't finding the common ancestor of the reorg.  For example we might have:

```mermaid
flowchart LR
 1 ---> 2a ---> 3a ---> 4a
 1 ---> 2b ---> 3b ---> 4b
```

The sequence of head events we'd get would be `1`, `2a`, `3a`, `4a` as the chain initially progresses, but then the reorg would just give a new head `4b`.  We don't have enough information to determine what the common ancestor is so we have to discard the entire cache.

it's really only the `else if` case below that works here - if the parent of the new head is in the cache and it's hash matches then we can rewind the cached blocks to the parent and push our new head.  The rewind is important and currently missing because we might reorg from `4a` to `2b` so the parent matches, but the chain has actually shortened and we need to remove some entries.

Reply:
In the above case of a 4a -> 4b reorg, are you suggesting that this should recursively fetch parent blocks until it finds a common ancestor (or exhausts the cache)?
In other words, upon receipt of 4b, fetch 3 (3b), then fetch 2 (2b), and see that 2b's parent is 1.
Should we set an upper bound to the amount of blocks we're willing to walk back?

Reply:
I don't think I'd bother walking back to find the common ancestor and just clear the whole cache in that situation. The worst case winds up back in the situation we have now where the cache misses and we have to fetch inline, but if we walked backwards we'd be doing that inline anyway so would be no better off. Since there are extremely few multiple block reorgs the vast majority of the time we'll hit the case where the parent of the new head is in the cache and can just build on that.

Reply:
I updated the logic to treat deeper reorgs such as the 4a->4b condition as a cache invalidation, and also to ensure that chain shortening is correctly applied.